src/transcoding.ts

```diff
@@ -0,0 +1,318 @@
+  const parts = field.split('.');
+  let value: RequestType | string | number | undefined = request;
+  for (const part of parts) {
+    if (typeof value !== 'object') {
+      return undefined;
+    }
+    value = (value as RequestType)[part] as RequestType;
+  }
+  if (typeof value === 'object' && !Array.isArray(value) && value !== null) {
+    return undefined;
+  }
+  return value;
+}
+
+export function deepCopy(request: RequestType): RequestType {
+  if (typeof request !== 'object' || request === null) {
+    return request;
+  }
+  const copy = Object.assign({}, request);
+  for (const key in copy) {
+    if (Array.isArray(copy[key])) {
+      copy[key] = (copy[key] as RequestType[]).map(deepCopy);
+    } else if (typeof copy[key] === 'object' && copy[key] !== null) {
+      copy[key] = deepCopy(copy[key] as RequestType);
+    }
+  }
+  return copy;
+}
+
+export function deleteField(request: RequestType, field: string): void {
+  const parts = field.split('.');
+  while (parts.length > 1) {
+    if (typeof request !== 'object') {
+      return;
+    }
+    const part = parts.shift() as string;
+    request = request[part] as RequestType;
+  }
+  const part = parts.shift() as string;
+  if (typeof request !== 'object') {
+    return;
+  }
+  delete request[part];
+}
+
+export function buildQueryStringComponents(
+  request: RequestType,
+  prefix = ''
+): string[] {
+  const resultList = [];
+  for (const key in request) {
+    if (Array.isArray(request[key])) {
+      for (const value of request[key] as RequestType[]) {
+        resultList.push(
+          `${prefix}${encodeWithoutSlashes(key)}=${encodeWithoutSlashes(
+            value.toString()
+          )}`
+        );
+      }
+    } else if (typeof request[key] === 'object' && request[key] !== null) {
+      resultList.push(
+        ...buildQueryStringComponents(request[key] as RequestType, `${key}.`)
+      );
+    } else {
+      resultList.push(
+        `${prefix}${encodeWithoutSlashes(key)}=${encodeWithoutSlashes(
+          request[key].toString()
+        )}`
+      );
+    }
+  }
+  return resultList;
+}
+
+export function encodeWithSlashes(str: string): string {
+  return str
+    .split('')
+    .map(c => (c.match(/[-_.~0-9a-zA-Z]/) ? c : encodeURIComponent(c)))
+    .join('');
+}
+
+export function encodeWithoutSlashes(str: string): string {
+  return str
+    .split('')
+    .map(c => (c.match(/[-_.~0-9a-zA-Z/]/) ? c : encodeURIComponent(c)))
+    .join('');
+}
+
+function escapeRegExp(str: string) {
+  return str.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
+}
+
+export function applyPattern(
+  pattern: string,
+  fieldValue: string
+): string | undefined {
+  if (!pattern || pattern === '*') {
+    return encodeWithSlashes(fieldValue);
+  }
+
+  if (!pattern.includes('*') && pattern !== fieldValue) {
+    return undefined;
+  }
+
+  // since we're converting the pattern to a regex, make necessary precautions:
+  const regex = new RegExp(
```

Comment:
did you check Node.js' url helpers, I wonder if there are any built in helpers we can use.

Reply:
Really disappointed but... https://github.com/benjamingr/RegExp.escape/issues/37

Reply:
I would love to have Perl's `quotemeta` here but... no.